eos-image-boot-dm-setup

```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
+# Copyright (C) 2016 Endless Mobile, Inc.
+# Licensed under the GPLv2
+#
+# When booting from an image file hosted on a filesystem, this program maps the
+# host filesystem to a dm-device with a hole where the image file lives.
+
+for i in $(</proc/cmdline); do
+  case $i in
+  endless.image.device=*)
+    host_device=${i//endless.image.device=}
+    ;;
+  endless.image.path=*)
+    image_path=${i//endless.image.path=}
+    ;;
+  esac
+done
+
+dm_roflag="--readonly"
+
+[ -z "${host_device}" -o -z "${image_path}" ] && exit 1
+
+case "${host_device}" in
+PARTUUID=*)
+  host_device=/dev/disk/by-partuuid/${host_device#PARTUUID=}
+  ;;
+UUID=*)
+  host_device=/dev/disk/by-uuid/${host_device#UUID=}
+  ;;
+esac
+
+host_device_size=$(blockdev --getsz ${host_device})
+if [ $? != 0 ]; then
+  echo "Failed to read ${host_device} size"
+  exit 1
+fi
+
+fstype=$(blkid -o value -s TYPE "${host_device}")
+if [ $? != 0 ]; then
+  echo "Failed to detect filesystem type on ${host_device}"
+  exit 1
+fi
+
+case "${fstype}" in
+exfat)
+  extents=$(dumpexfat -f "${image_path}" "${host_device}")
+  ;;
+ntfs)
+  extents=$(ntfsextents "${host_device}" "${image_path}")
+  ;;
+*)
+  echo "Unsupported filesystem ${fstype} on ${host_device}"
+  exit 1
+esac
+
+if [ $? != 0 ]; then
+  echo "Failed to lookup ${image_path} on ${host_device} (${fstype})"
+  exit 1
+fi
+
+# Cleanup the device mapper table file from previous runs
+rm -f /tmp/dmtable
+
+# Sort the file extents we create the holes in the correct places
+extents=$(sort ${extents})
```

Comment:
Yeah, you are right, I did some quick testing on files and stopped thinking before writing the actual code.